src/AMDGPUOffload.cpp

```diff
@@ -0,0 +1,679 @@
+enum {
+    EV_CURRENT = 1,
+};
+
+enum {
+    EM_AMDGPU = 224,
+};
+
+// http://www.llvm.org/docs/AMDGPUUsage.html#code-object
+enum {
+    EF_AMDGPU_MACH_AMDGCN_GFX801 = 0x028, //       gfx801
+    EF_AMDGPU_MACH_AMDGCN_GFX802 = 0x029, //       gfx802
+    EF_AMDGPU_MACH_AMDGCN_GFX803 = 0x02a, //       gfx803
+    EF_AMDGPU_MACH_AMDGCN_GFX810 = 0x02b, //       gfx810
+    EF_AMDGPU_MACH_AMDGCN_GFX900 = 0x02c, //       gfx900
+    EF_AMDGPU_MACH_AMDGCN_GFX902 = 0x02d, //       gfx902
+    EF_AMDGPU_XNACK = 0x100
+};
+
+// https://github.com/llvm-mirror/llvm/blob/master/include/llvm/BinaryFormat/ELFRelocs/AMDGPU.def
+enum {
+    R_AMDGPU_NONE = 0,
+    R_AMDGPU_ABS32_LO = 1,
+    R_AMDGPU_ABS32_HI = 2,
+    R_AMDGPU_ABS64 = 3,
+    R_AMDGPU_REL32 = 4,
+    R_AMDGPU_REL64 = 5,
+    R_AMDGPU_ABS32 = 6,
+    R_AMDGPU_GOTPCREL = 7,
+    R_AMDGPU_GOTPCREL32_LO = 8,
+    R_AMDGPU_GOTPCREL32_HI = 9,
+    R_AMDGPU_REL32_LO = 10,
+    R_AMDGPU_REL32_HI = 11,
+    R_AMDGPU_RELATIVE64 = 13,
+};
+
+std::string hex(uint32_t x) {
+    char buffer[100];
+    snprintf(buffer, sizeof(buffer), "0x%08x", x);
+    return buffer;
+}
+
+std::string section_type_string(Section::Type type) {
+    switch(type) {
+    case Section::SHT_NULL: return "SHT_NULL";
+    case Section::SHT_PROGBITS: return "SHT_PROGBITS";
+    case Section::SHT_SYMTAB: return "SHT_SYMTAB";
+    case Section::SHT_STRTAB: return "SHT_STRTAB";
+    case Section::SHT_RELA: return "SHT_RELA";
+    case Section::SHT_HASH: return "SHT_HASH";
+    case Section::SHT_DYNAMIC: return "SHT_DYNAMIC";
+    case Section::SHT_NOTE: return "SHT_NOTE";
+    case Section::SHT_NOBITS: return "SHT_NOBITS";
+    case Section::SHT_REL: return "SHT_REL";
+    case Section::SHT_SHLIB: return "SHT_SHLIB";
+    case Section::SHT_DYNSYM: return "SHT_DYNSYM";
+    case Section::SHT_LOPROC: return "SHT_LOPROC";
+    case Section::SHT_HIPROC: return "SHT_HIPROC";
+    case Section::SHT_LOUSER: return "SHT_LOUSER";
+    case Section::SHT_HIUSER: return "SHT_HIUSER";
+    default:
+        return "UNKNOWN TYPE";
+    }
+}
+std::string print_sections(const Object &obj) {
+    std::ostringstream oss;
+    if (obj.sections_size() == 0) {
+        oss << "No sections in object\n";
+        return oss.str();
+    }
+    for (const Section &s: obj.sections()) {
+        oss << s.get_name() << ", Type = " << section_type_string(s.get_type()) << ", Size = " << hex(s.get_size()) << ", Alignment = " << s.get_alignment() << "\n";
+    }
+    return oss.str();
+}
+
+void do_reloc_64(char *addr, uintptr_t val) {
+    *((uint64_t*)addr) = (uint64_t)val;
+}
+
+void do_reloc_32(char *addr, uintptr_t val) {
+    internal_assert(val <= 0xffffffff);
+    *((uint32_t*)addr) = val;
+}
+
+void do_relocation(uint32_t fixup_offset, char *fixup_addr, uint32_t type,
+                   const Symbol *sym, uint32_t sym_offset, int32_t addend,
+                   Elf::Section &got) {
```

Comment:
Add got_offset as an additional argument and use to initialize the GOT variable used in the R_AMDGPU_GOT* relocations.

Offset arguments need to be uint64_t as addresses are 64 bit. (I guess could remain 32 bit as this code sets the ELF load address as 0 and is limited to 4GiB code objects which is likely ok.)